Add support for systemair-savecair climate system

perara opened the pull request:
## Description:


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#6951

## Example entry for `configuration.yaml` (if applicable):
```yaml
climate:
  - platform: systemair_savecair
    name: System AIR
    iam_id: IAM_XXXXXXXXXXX
    password: xxxxxxxxxx
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

perara committed:
Added support for systemair-savecair climate system

perara committed:
Updated linting for systemair_savecair component

hwikene commented:
This is awesome. I work at Systemair, let me know if you need any help.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+    vol.Required(CONF_DEVICE): cv.string,
+    vol.Required(CONF_IAM_ID): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_SCAN_INTERVAL, default=SCAN_INTERVAL): cv.time_period,
```
Remove scan interval. It's already present in the base platform config schema.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+                               discovery_info=None):
+    """Set up the SystemAIR platform."""
+    name = config.get(CONF_NAME)
+    device = config.get(CONF_DEVICE)
```
Use `dict[key]` for required config keys.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+SystemAIR platform for VTR 300 Ventilation Unit.
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/savecair/
```
The url format is wrong. Please look at other platforms for the correct format.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+
+        # Create savecair client
+        self._client = SavecairClient(iam_id, password)
+        self._client.update_cb.append(self.update_callback)
```
Please move callback registration to entity coroutine method `async_added_to_hass`. Otherwise if an update comes in before the entity has been added to home assistant there will be an error.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+        self.last_operation_mode = None
+
+        self._fan_list = ['Off', 'Low', 'Normal', 'High']
+        self._operation_list = ['Auto', 'Manual', 'Crowded',
```
Operation list must only contain states found in the base climate component. Use two dicts to go from home assistant state to device mode.

perara replied:
I've been thinking about this. So basicly, SystemAIR delivers ventilation units for ventilation in homes. Thes have several operation modes, such as "Away, Holiday ...etc" 

The reason I put this component under climate is that it also supports Heating and Cooling in addition to. I do believe that it fits well here because its really just a climate device. I'm not sure on how to approach this issue. Should I suggest additional operations added to HA, or is there a way to show the SystemAIR operations in frontend while using HA operations in the backend:
```
self._operation_list = [STATE_AUTO, STATE_MANUAL, STATE_AUTO, STATE_AUTO, STATE_AUTO, STATE_AUTO, STATE_AUTO]
self._operation_list_map = ['Auto', 'Manual', 'Crowded','Refresh', 'Fireplace', 'Away', 'Holiday']
 ```

perara replied:
@MartinHjelmare, do you have any input here? :)

MartinHjelmare replied:
There's no way around using the modes in the base climate component. You're welcome to propose a change, but I can't promise that it will be accepted and I don't know how long that could take. We currently have an open issue on climate component restructuring in our architecture repo. Please start by reading that.

perara replied:
@MartinHjelmare Maybe, these devices dont belong in the climate component then? These are primarily ventilation units, but there exists about 20 different units. There is 2 ways of interfacing using savecair module or modbus. Would it be possible to have systemair as a base component, or maybe a ventilation base component?

I'm really not sure about what is the most fitting, but i'm quite sure that systemair will be a increasingly used device in newer homes.

MartinHjelmare replied:
Why not start by implementing auto, manual and away, which all already exist in the climate component?

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+        self._operation_list = ['Auto', 'Manual', 'Crowded',
+                                'Refresh', 'Fireplace', 'Away', 'Holiday']
+
+    def update_callback(self):
```
Will this be called from a thread or from the event loop? Use our async api for async safe calls and our sync api for non async calls. Context is key.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+    async def async_update(self):
+        """Retrieve latest state."""
+        await self._client.update_sensors()
+        self.schedule_update_ha_state()
```
Use async api, ie `async_schedule_update_ha_state`.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+        self.schedule_update_ha_state()
+
+    @property
+    def state_attributes(self):
```
This should be `device_state_attributes`.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+    @property
+    def state_attributes(self):
+        """Update state attributes with additional attributes."""
+        attributes = super().state_attributes
```
Remove this.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+            ATTR_SUPPLY_AIR_FAN_SPEED: self.extract_air_fan_speed,
+            ATTR_EXTRACT_AIR_FAN_SPEED: self.supply_air_fan_speed
+        })
+        attributes.update(self._client.data)
```
What data are we adding here?

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return True
```
This is the default.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+    @property
+    def supply_air_fan_speed(self):
+        """Return the supply air fan speed."""
+        return self._client.data["digital_input_tacho_saf_value"] \
```
Please rewrite all if else statements that span multiple lines to separate statements. Prefer guard clause when possible.
```py
if "digital_input_tacho_saf_value" not in self._client.data:
    return None
return self._client.data["digital_input_tacho_saf_value"]
```

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+        if "main_user_mode" not in self._client.data:
+            return None
+
+        return self._operation_list[
```
Only home assistant states are allowed to be returned here.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+        """Return the supported step of target temperature."""
+        return 1
+
+    def set_temperature(self, **kwargs):
```
Change to async version if we want to use async functions inside.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+
+        self.hass.async_create_task(self._client.set_temperature(temperature))
+
+    def set_fan_mode(self, fan_mode):
```
Same as above.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+        elif fan_mode == "Maximum":
+            self.hass.async_create_task(self._client.set_fan_high())
+
+    def set_operation_mode(self, operation_mode):
```
Same as above.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+
+    def set_operation_mode(self, operation_mode):
+        """Set the operation mode."""
+        if self.last_operation_mode == "Fireplace":
```
Use one of the dicts here that map home assistant state to device mode.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+
+        self.last_operation_mode = operation_mode
+
+    def turn_on(self):
```
Same as above.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,249 @@
+        """Turn on the climate device."""
+        self.hass.async_create_task(self._client.set_fan_high())
+
+    def turn_off(self):
```
Same as above.

MartinHjelmare commented:
Is polling necessary when we have the state update callback?

perara commented:
@MartinHjelmare Thanks for the feedback. Will update accordingly. Heres a few questions:
@MartinHjelmare Thanks for the feedback. Will update accordingly. Heres a few questions:

1. Is polling necessary when we have the state update callback?
Depends, the reason I added this is that it did not receieve a state update on HA boot. I could probably fix this so that we dont use polling tho.

2. Regarding operation list. I dont find it natural to use the base states, they just dont fit the modes found in the systemair system. I could off course map them "randomly" so lets say Crowded => STATE_DRY ...etc, but would not this make the states hard to track internally?
```
STATE_HEAT = 'heat'
STATE_COOL = 'cool'
STATE_IDLE = 'idle'
STATE_AUTO = 'auto'
STATE_MANUAL = 'manual'
STATE_DRY = 'dry'
STATE_FAN_ONLY = 'fan_only'
STATE_ECO = 'eco'
```

3. ` attributes.update(self._client.data)` just adds everything from the systemair object. this makes the object abit "dirty", but i thought then atleast the user has the ability to use all data available. Should this be removed?

MartinHjelmare commented:
1. Call `async_add_entities` like this `async_add_entities([climate], True)` to have `async_update` be called during entity addition. Then set `should_poll` to `False`.
2. We use a strict entity model. Only states present in the base climate component may be used for operation mode. Please try to map the device modes as best as possible. New climate states can only be added after approval in an issue in our architecture repo: https://developers.home-assistant.io/docs/en/entity_index.html#changing-the-entity-model
3. Can you list the attributes of the systemair object? They may be ok to be added, but we have some restrictions, so it depends on what they are. State attributes should generally be used for attributes that concern the state of the entity. We do now allow to add eg firmware version or other similar static information.

perara commented:
@MartinHjelmare I could clean this up abit, but generally this is what is retrieved from the API
```
current_temperature: 24.6
min_temp: 7
max_temp: 35
temperature: 20
target_temp_step: 1
fan_mode: High
fan_list: Off,Low,Normal,High
operation_mode: Manual
operation_list: Auto,Manual,Crowded,Refresh,Fireplace,Away,Holiday
supply_air_temperature: 19.3
supply_air_fan_speed: 2520
extract_air_fan_speed: 2520
alarm_emt_state: inactive
rh_sensor: 47
user_mode_fireplace_supply: 5
alarm_extra_controller_state: inactive
control_regulation_airflow_manual_fan_stop: true
alarm_oat_state: inactive
function_active_heat_recovery: true
function_active_moisture_transfer: false
alarm_external_stop_state: inactive
function_active_service_user_lock: false
control_regulation_demand_co2_status: false
user_mode_refresh_extract: 5
function_active_defrosting: false
function_active_cooker_hood: false
alarm_pdm_rhs_state: inactive
alarm_eat_state: inactive
alarm_ect_state: inactive
alarm_eft_state: inactive
alarm_manual_mode_state: inactive
main_user_mode: 1
fan_log_request_reset: 30227
alarm_overheat_temperature_state: inactive
alarm_co2_state: inactive
user_lock: false
eco_mode: false
alarm_filter_state: inactive
week_schedule_any_defined: true
alarm_oht_state: inactive
week_schedule_lock: true
alarm_sat_state: inactive
control_regulation_demand_rh_status: false
alarm_eaf_rpm_state: inactive
function_active_cooling: false
function_active_cooling_recovery: false
supply_air_temp: 193
main_airflow: 4
control_regulation_airflow_type: rpm
outdoor_air_temp: 134
alarm_frost_prot_state: inactive
alarm_saf_ctrl_state: inactive
user_mode_holiday_extract: 1
alarm_rgs_state: inactive
function_active_secondary_air: false
user_mode_crowded_extract: 5
user_mode_holiday_supply: 1
digital_input_tacho_saf_value: 2520
alarm_rat_state: inactive
alarm_low_sat_state: inactive
main_temperature_offset: 200
user_mode_crowded_supply: 5
user_mode_fireplace_extract: 1
user_mode_auto_supply: 4
function_active_heating: false
alarm_fpt_state: inactive
main_iaq: economic
user_mode_refresh_supply: 5
alarm_rh_state: inactive
alarm_secondary_air_state: inactive
filter_lock: true
digital_input_tacho_eaf_value: 2520
components_filter_time_left: 18594015
alarm_bys_state: inactive
alarm_eaf_ctrl_state: inactive
alarm_defrosting_state: inactive
alarm_manual_fan_stop_state: inactive
pdm_input_temp_value: 246
function_active_heater_cooldown: false
function_active_free_cooling: false
alarm_saf_rpm_state: inactive
user_mode_away_supply: 1
function_active_vacuum_cleaner: false
control_regulation_airflow_unit_man_perc: null
control_regulation_temp_unit: celsius
alarm_pdm_eat_state: inactive
user_mode_away_extract: 1
friendly_name: System AIR
supported_features: 4289
```

perara commented:
@hwikene See https://github.com/perara/python-systemair-savecair/issues/1

perara commented:
> How often is `components_filter_time_left` time changed? Is it changed on every state update? We don't want to spam the state machine just because time has passed. It's ok if it's updated infrequently by the device. Generally we prefer absolute time points in state attributes.

I do believe this changes in real-time, meaning that it would be different each poll in HA. Such information could however, be beneficial for automations. Do you want me to remove such data?

MartinHjelmare commented:
Yes, then please remove it. It it's possible to replace it with a corresponding absolute time point, that would be ok.

hwikene reviewed (commented):
I think ```components_filter_time_left```should be an option. It is great for automations or send a notification when its time to change filters. The option on the unit is 6, 9, and 12 months for each change.

perara committed:
Updated for climate support

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,311 @@
+DEFAULT_NAME = "SystemAIR"
+DEFAULT_DEVICE = "vtr_300"
+
+SIGNAL_SYSTEMAIR_UPDATE = "systemair_update"
```
This isn't used.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,311 @@
+    def __init__(self, hass, name, unit, iam_id, password):
+        """Construct the SystemAIR Climate Device."""
+        self._name = name
+        self._unit = unit
```
This isn't used.

perara committed:
Updated systemair according to feedback

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,269 @@
+    @property
+    def current_operation(self):
+        """Return current operation ie. heat, cool, idle."""
+        return self._client.get_current_operation()
```
This should return one of the climate base component states. If possible create a persistent dict that maps client mode to home assistant state lookup the correct state here.

Create the persistent dict in init, or if we don't need to import anything from the client library to create the dict, we can put it at the module level.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,269 @@
+
+    async def async_set_operation_mode(self, operation_mode):
+        """Set the operation mode."""
+        await self._client.set_operation_mode(operation_mode)
```
Operation mode is one of the home assistant climate base component states. Use a dict to convert here too.

perara committed:
Fixed linting
Added operation modes correctly

perara committed:
Removed hard-coded path to external library...

perara committed:
Updated requirements for systemair-savecair

perara committed:
Added docstring

perara committed:
Updated according to requested changes in #17608

perara committed:
Flake8 fix

perara committed:
Flake8 fix

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,291 @@
+    @property
+    def operation_list(self):
+        """Return the list of available operation modes."""
+        return self._operation_list
```
Return the constant here directly.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,291 @@
+        if self.current_operation != STATE_MANUAL:
+            return None
+
+        return self._fan_list
```
Return the constant here directly.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,291 @@
+        """Construct the SystemAIR Climate Device."""
+        self._name = name
+
+        self._support_flags = SUPPORT_TARGET_TEMPERATURE
```
Looks like we can extract this into a constant at the module level too.

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,291 @@
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        return self._support_flags
```
Return the constant here.

perara committed:
Updated according to reviewer specs:
SUPPORT_FLAGS now at module level
FAN_LIST at module level
OPERATION_LIST at module level

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,288 @@
+    2: STATE_MANUAL,
+}
+
+HA_OPERATION_MODES = {
```
```py
HA_OPERATION_MODES = {
    val: key for key, val in SAVECAIR_OPERATION_MODES.items()
}
```

MartinHjelmare reviewed (commented):
Last comment. Looks good!

perara committed:
Updated according to reviewer feedback

MartinHjelmare reviewed (approved):
Great! Can be merged when build passes.

perara committed:
Updated to accommodate coverage issues

perara committed:
Fixed issue where fan_list was None

perara committed:
Version bump

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,286 @@
+    def fan_list(self):
+        """Return the list of available fan modes."""
+        if self.current_operation != STATE_MANUAL:
+            return []
```
Why return an empty list?

perara replied:
I received feedback that when fan_list returned None, HA spammed the log with "fan_list does not exists". They utilized an input_select, and read the fan_list. I figured one solution would be to return a empty list to remedy this. Thoughts?

perara replied:
I can off course revert this, if it does not comply.

MartinHjelmare replied:
If the entity doesn't support `fan_list` it should return `None` for that property (and not support that feature, but features can't change during the entity lifetime either).

Templates always need to handle non existing attributes. So I don't think we should change the property for that reason.

perara committed:
Revert back to None on Fan_list

MartinHjelmare commented on homeassistant/components/climate/systemair_savecair.py:
```diff
@@ -0,0 +1,286 @@
+
+    # Create savecair client
+    from systemair.savecair import SaveCairClient
+    client = SaveCairClient(iam_id, password, loop=hass.loop)
```
We should check that the credentials are correct before continuing with set up. If they are bad, log error and return.

balloob commented:
It looks like the dependency is changing the operation modes in Home Assistant to work around our requirement for operation modes.

https://github.com/perara/python-systemair-savecair/blob/d906a0f49e2ddec7e359d7f43b107f62f63dadac/systemair/savecair/client.py#L58-L74

![image](https://user-images.githubusercontent.com/1444314/48552434-e6f20200-e8d8-11e8-8bdd-a6ebcaf2fa8d.png)


Not only is not not allowed, it violates our rules and disrespects us, the project and all the work we do.

Closing this PR and blocking @perara from interacting with our community for a week.

perara commented:
Hi, 
As mentioned privately, this was not intended for the publicly available home-assistant component.
Although I found it disappointing that I was not allowed to comment before the pr close and got banned from contributing (lol), I'm more interested in finalizing this feature to the benefit of others.
Would you allow me to correctly upload the public version of the library and finalize the work?

prilly-dev commented:
I would realy like to see this integration finished. Using Modbus with this unit is a very time consuming and it leav out the possibility to use the cloud solution.

frenck commented:
@prilly-dev This PR has been closed for 2 years now. Please don't bring up these PRs. If you have an issue, please raise an issue. In case of a feature request, please use our community forum.

Thanks 👍
